middleman_protocol/middleman_protocol/message.py

```diff
@@ -0,0 +1,302 @@
+    - parsing serialized bytes payload back to original format on deserialization,
+    - validation of original payload before serialization.
+
+    """
+
+    __slots__ = [
+        'payload',
+        'payload_type',
+        'request_id',
+    ]
+
+    def __init__(self, payload: Any, request_id: int) -> None:
+        self._validate_request_id(request_id)
+        self.request_id = request_id
+
+        self._validate_payload(payload)
+        self.payload = self._serialize_payload(payload)
+
+    @classmethod
+    @abstractclassmethod
+    def _deserialize_payload(cls, payload: bytes):
+        pass
+
+    @abstractmethod
+    def _serialize_payload(self, payload: Any):
+        pass
+
+    @abstractmethod
+    def _validate_payload(self, payload: Any):
+        pass
+
+    @classmethod
+    def get_frame_format(cls) -> Struct:
+        """ Returns Struct object containing frame structure. """
+        return Struct(
+            frame_signature=Bytes(65),
+            signed_part_of_the_frame=Struct(
+                request_id=BytesInteger(4),
+                payload_type=Enum(Byte, PayloadType),
+                payload_length=BytesInteger(4),
+                payload=GreedyBytes,
+            ),
+        )
+
+    @classmethod
+    def add_frame_separator(cls, raw_frame: bytes) -> bytes:
+        """ Adds frame separator to raw frame. """
+        assert isinstance(raw_frame, bytes)
+
+        return raw_frame + FRAME_SEPARATOR
+
+    @classmethod
+    def remove_frame_separator(cls, raw_message: bytes) -> bytes:
+        """ Remove frame separator from raw frame. """
+        assert isinstance(raw_message, bytes)
+        assert FRAME_SEPARATOR in raw_message
+
+        frame_separator_index = raw_message.index(FRAME_SEPARATOR)
+
+        assert len(raw_message) == frame_separator_index + FRAME_SEPARATOR_LENGTH
+
+        return raw_message[:frame_separator_index]
+
+    @classmethod
+    def deserialize(cls, raw_message: bytes, public_key: bytes, is_frame_separator_expected: bool=True) -> Any:
+        """
+        Parses and validates received message in bytes.
+        Returns original payload.
+        """
+
+        # Remove frame separator from raw message if it is expected
+        if is_frame_separator_expected is True:
+            raw_message = cls.remove_frame_separator(raw_message)
+
+        # Parse frame
+        frame_format = cls.get_frame_format()
+        frame = frame_format.parse(raw_message)
+
+        raw_signed_part_of_the_frame = frame_format.signed_part_of_the_frame.build(
+            frame.signed_part_of_the_frame
+        )
+
+        # Validate
+        cls._validate_signature(raw_signed_part_of_the_frame, frame.frame_signature, public_key)
+        cls._validate_length(frame)
+
+        # Get class related to current payload type
+        message_class = PAYLOAD_TYPE_TO_MIDDLEMAN_MESSAGE_CLASS[
+            PayloadType[str(frame.signed_part_of_the_frame.payload_type)]
+        ]
```

Comment:
What if there's no such class? This should not result in a crash. It should be treated as an invalid message.

I think you're not validating the payload type. Unless it's done in a different way, not with a `_validate_xxx()` method?